server.py

```diff
@@ -23,53 +28,137 @@
 def hello():
     return 'hello!'
 
-# Endpoint for PUT method
-@app.route('/write', methods=['PUT'])
+@app.route('/redirect')
+def redirect_endpoint():
+    return redirect('http://localhost:5000/read?uuid=xxx', code=302)
+
+# Endpoint for write method
+@app.route('/write', methods=['POST'])
 def write_file():
     file = request.files['file']
     if file:
         filename = secure_filename(file.filename)
         file_uuid = str(uuid.uuid4())
-        # return str(os.path.join(app.config['UPLOAD_FOLDER'], file_uuid)), 200
         file.save(os.path.join(app.config['UPLOAD_FOLDER'], file_uuid))
+        metadata = metadata_manager.MetadataManager()
+        metadata.update_file_stored(file_uuid, app.config['HOST'])
+        metadata.close_connection()
         return file_uuid, 201
     return 'Write Failed', 500
 
-# Endpoint for GET method
+# Endpoint for read method
 @app.route('/read', methods=['GET'])
 def read_file():
-    sys.stdout.flush()
     filename = request.args.get('uuid')
-    return send_from_directory(UPLOAD_FOLDER, secure_filename(filename))
+    file_path = UPLOAD_FOLDER + '/' + secure_filename(filename)
+    metadata = metadata_manager.MetadataManager()
+    if (os.path.exists(file_path)):
+        return send_from_directory(UPLOAD_FOLDER, secure_filename(filename))
+
+    redirect_address = metadata.lookup_file(filename, app.config['HOST'])
+    if (redirect_address is not None):
+        url = 'http://%s/read?%s' % (redirect_address, urllib.urlencode({ 'uuid': filename }))
+        return redirect(url, code=302)
+
+    other_servers = metadata.get_all_server(app.config['HOST'])
+    if (len(other_servers) > 0):
+        for server in other_servers:
+            url = 'http://%s/file_exists?%s' % (server, urllib.urlencode({ 'uuid': filename }))
+            lookup_request = requests.get(url)
+            if (lookup_request.status_code == 200):
+                redirection_url = 'http://%s/read?%s' % (server, urllib.urlencode({ 'uuid': filename }))
+                metadata.update_file_stored(filename, server)
+                metadata.close_connection()
+                return redirect(redirection_url, code=302)
```

Comment:
I think we agree that it's already a miss on the closest server to the
client, so returning anywhere is fine. no?

On Mon, Mar 2, 2015 at 1:06 PM, Chris Baik <notifications@github.com> wrote:

> In server.py
> <https://github.com/paivaspol/EECS591/pull/3#discussion_r25621438>:
>
> > +
> > +    redirect_address = metadata.lookup_file(filename, app.config['HOST'])
> > +    if (redirect_address is not None):
> > +        url = 'http://%s/read?%s' % (redirect_address, urllib.urlencode({ 'uuid': filename }))
> > +        return redirect(url, code=302)
> > +
> > +    other_servers = metadata.get_all_server(app.config['HOST'])
> > +    if (len(other_servers) > 0):
> > +        for server in other_servers:
> > +            url = 'http://%s/file_exists?%s' % (server, urllib.urlencode({ 'uuid': filename }))
> > +            lookup_request = requests.get(url)
> > +            if (lookup_request.status_code == 200):
> > +                redirection_url = 'http://%s/read?%s' % (server, urllib.urlencode({ 'uuid': filename }))
> > +                metadata.update_file_stored(filename, server)
> > +                metadata.close_connection()
> > +                return redirect(redirection_url, code=302)
>
> I see. Hmmm.. Is it correct to return the first one we hit? What if that's
> the farthest away from the client?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/paivaspol/EECS591/pull/3/files#r25621438>.
>